### PR TITLE
Add operator commands

### PR DESCRIPTION
**Pull request by Techassi**
This PR adds all operator related commands.

### CODE REVIEW

File: bins/stackablectl/src/cmds/operator.rs [hunk 1]
```diff
@@ -72,18 +93,37 @@ Use \"stackablectl operator describe <OPERATOR>\" to get available versions for
...
     )]
-    cluster: ClusterType,
+    cluster_type: ClusterType,
 
     /// Name of the local cluster
     #[arg(long, default_value = DEFAULT_LOCAL_CLUSTER_NAME)]
     cluster_name: String,
+
+    /// Number of total nodes in the local cluster
+    #[arg(long, default_value_t = 3)]
```
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs:110]: Currently this describes the **total** number of cluster nodes. The lowest this can go is 2 as the cluster needs one control-plane node and one worker node. But I can change the default value to `2`.

[the remark this answers is not in the document]
Comment (sbernauer) [reply bins/stackablectl/src/cmds/operator.rs:110]: Oh sorry, 2 than please

File: bins/stackablectl/src/cli.rs [hunk 2]
```diff
@@ -44,7 +46,7 @@ Cached files are saved at '$XDG_CACHE_HOME/stackablectl', which is usually
     pub offline: bool,
 
     /// Namespace in the cluster used to deploy the products and operators
-    #[arg(short, long, default_value = DEFAULT_STACKABLE_NAMESPACE)]
+    #[arg(short, long, default_value = DEFAULT_NAMESPACE)]
     pub namespace: String,
```
Comment (sbernauer) [inline bins/stackablectl/src/cli.rs:50]: Thinking of https://github.com/stackabletech/stackablectl/issues/259 I would prefer calling the option `--operators-namespace` with the default value of `DEFAULT_OPERATORS_NAMESPACE`, being `stackable-operators` (or `default` for now).
Going the final way with `stackable-operators` and creating the namespace on demand would be awesome!

File: bins/stackablectl/src/cmds/operator.rs [hunk 3]
```diff
@@ -100,42 +140,346 @@ pub struct OperatorInstalledArgs {
...
+    ClusterError { source: ClusterError },
+
+    #[snafu(display("semver parse error: {source}"))]
+    SemVerParseError { source: semver::Error },
+
+    #[snafu(display("yaml error: {source}"))]
+    YamlError { source: serde_yaml::Error },
+
+    #[snafu(display("json error: {source}"))]
+    JsonError { source: serde_json::Error },
```
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs:163]: Okay what is the way forward when using Snafu? Does Snafu automatically include the underlying error in the `Display`?

[the remark this answers is not in the document]
Comment (nightkr) [reply bins/stackablectl/src/cmds/operator.rs:163]: No, it's the job of whoever is printing the error traverse the `source()` chain. For example, see https://docs.rs/eyre/latest/eyre/struct.Report.html#display-representations.

File: bins/stackablectl/src/cmds/operator.rs [hunk 4]
```diff
@@ -100,42 +140,346 @@ pub struct OperatorInstalledArgs {
...
+                .load_preset(NOTHING)
+                .add_row(vec!["OPERATOR", &args.operator_name.to_string()])
+                .add_row(vec!["STABLE VERSIONS", stable_versions_string.as_str()])
+                .add_row(vec!["TEST VERSIONS", test_versions_string.as_str()])
+                .add_row(vec!["DEV VERSIONS", dev_versions_string.as_str()]);
+
+            Ok(table.to_string())
+        }
+        OutputType::Json => todo!(),
+        OutputType::Yaml => todo!(),
```
[the remark this answers is not in the document]
Comment (nightkr) [reply bins/stackablectl/src/cmds/operator.rs:268]: I'm not saying we should get rid of the YAML/JSON outputs, just that we ideally shouldn't have to implement this separately for each command.

File: bins/stackablectl/src/cli.rs (hunk 2 above)
Comment (sbernauer) [reply bins/stackablectl/src/cli.rs:50]: Works for me as well. I was just thinking "well there are multiple operators running in this namespace".
TLDR: Do whatever you desire :)

File: bins/stackablectl/src/cmds/operator.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nightkr) [reply bins/stackablectl/src/cmds/operator.rs:110]: It's valid to run a single node as both a control plane and worker (although it's definitely not recommended for production setups).

[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs]: Changed in 5b0c371.

File: bins/stackablectl/src/cmds/operator.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs]: Changed in 5b0c371.

File: bins/stackablectl/src/cmds/operator.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs]: Added in 5b0c371.

File: bins/stackablectl/src/cli.rs (hunk 2 above)
Comment (Techassi) [reply bins/stackablectl/src/cli.rs]: Changed in 2bbd6f6.

File: src/cluster/kind/mod.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            namespace: namespace.unwrap_or(DEFAULT_STACKABLE_NAMESPACE.into()),
+            name: name.unwrap_or(DEFAULT_LOCAL_CLUSTER_NAME.into()),
+            cp_node_count,
+            node_count,
+        }
+    }
+
+    /// Create a new local cluster by calling the kind binary.
+    #[instrument]
+    pub fn create(&self) -> Result<(), ClusterError> {
```
Comment (nightkr) [inline src/cluster/kind/mod.rs:65]: Arguably this should be async (using `tokio::process` rather than `std::process`).

File: src/helm.rs [hunk 6]
```diff
@@ -18,6 +19,22 @@ pub struct HelmRelease {
...
 }
 
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct HelmChart {
+    pub release_name: String,
+    pub name: String,
+    pub repo: HelmChartRepo,
+    pub version: String,
+    pub options: serde_yaml::Value,
```
Comment (nightkr) [inline src/helm.rs:29]:
```suggestion
    pub values: serde_yaml::Value,
```

Comment (nightkr) [reply src/helm.rs:29]: We might as well be consistent with how Helm names it.

File: src/helm.rs [hunk 7]
```diff
@@ -347,16 +429,39 @@ pub fn add_repo(repo_name: &str, repo_url: &str) -> Result<(), HelmError> {
...
+        .context(RequestSnafu {})?
+        .text()
+        .await
+        .context(RequestSnafu {})?;
+
+    serde_yaml::from_str(&index_file_content).context(YamlSnafu {})
+}
+
 /// Helper function to convert raw C string pointers to &str.
 fn ptr_to_str<'a>(ptr: *const i8) -> Result<&'a str, Utf8Error> {
```
Comment (nightkr) [inline src/helm.rs:456]:
```suggestion
unsafe fn ptr_to_str<'a>(ptr: *const i8) -> Result<&'a str, Utf8Error> {
```

File: bins/stackablectl/src/cmds/operator.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs]: Okay. That seems reasonable. I will update the variant name to `JsonOutputFormatError`.

File: src/helm.rs (hunk 7 above)
Comment (nightkr) [reply src/helm.rs:456]: It's unsafe to convert a raw pointer to a reference, since it's undefined behaviour unless you can guarantee that Rust's lifetime, null, and aliasing rules won't be violated for as long as the reference exists.

File: bins/stackablectl/src/cmds/operator.rs [hunk 8]
```diff
@@ -100,42 +140,346 @@ pub struct OperatorInstalledArgs {
...
         }
     }
 }
 
-fn list_cmd(args: &OperatorListArgs) -> Result<String, OperatorError> {
+#[instrument]
+async fn list_cmd(args: &OperatorListArgs, common_args: &Cli) -> Result<String, OperatorError> {
+    debug!("Listing operators");
+
+    // If the user only wnats to list installed operator, use this shortcut
```
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs:193]: Yeah I was taking inspiration from other tools like pnpm, see [here](https://pnpm.io/cli/list). I can remove the shortcut if needed.

File: bins/stackablectl/src/cli.rs (hunk 2 above)
Comment (Techassi) [reply bins/stackablectl/src/cli.rs]: I agree with @nightkr, `operator-namespace` flows indeed a little better. Updated in 761a13e.

File: src/kube.rs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        .context(KubeSnafu {})?;
+
+    for manifest in serde_yaml::Deserializer::from_str(manifests) {
+        let mut object = DynamicObject::deserialize(manifest).context(YamlSnafu {})?;
+        let object_type = object.types.as_ref().ok_or(
+            ObjectTypeSnafu {
+                msg: format!(
+                    "failed to deploy manifest because type of object {object:?} is not set"
+                ),
+            }
```
Comment (nightkr) [inline src/kube.rs:39]: This error formatting should be in the `#[snafu(display)]`

File: src/kube.rs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+            .build(),
+        )?;
+
+        let gvk = gvk_of_typemeta(object_type);
+        let (resource, capabilities) = discovery.resolve_gvk(&gvk).ok_or(
+            DiscoverySnafu {
+                msg: format!("failed to deploy manifest because GVK {gvk:?} cannot be resolved."),
+            }
```
Comment (nightkr) [inline src/kube.rs:47]: Same here

File: src/platform/release/spec.rs [hunk 11]
```diff
@@ -40,16 +40,20 @@ pub struct ReleaseSpec {
 
 impl ReleaseSpec {
     #[instrument(skip_all)]
-    pub fn install(
+    pub fn install<T>(
         &self,
         include_products: &[String],
         exclude_products: &[String],
-    ) -> Result<(), ReleaseInstallError> {
+        namespace: T,
```
Comment (nightkr) [inline src/platform/release/spec.rs:47]: Namespace could just be a `&str` here, since it's not persisted beyond the function anyway.

File: src/platform/stack/spec.rs [hunk 12]
```diff
@@ -67,7 +93,10 @@ pub struct StackSpecV2 {
 
 impl StackSpecV2 {
     #[instrument(skip_all)]
-    pub fn install(&self, release_list: ReleaseList) -> Result<(), StackError> {
+    pub fn install<T>(&self, release_list: ReleaseList, namespace: T) -> Result<(), StackError>
```
Comment (nightkr) [inline src/platform/stack/spec.rs:96]: Another pointless generic.

File: src/platform/stack/spec.rs [hunk 13]
```diff
@@ -77,41 +106,114 @@ impl StackSpecV2 {
...
 
         todo!()
     }
 
     #[instrument(skip_all)]
-    pub fn install_stack_manifests(&self, parameters: &[String]) -> Result<(), StackError> {
+    pub async fn install_stack_manifests<T>(
+        &self,
+        parameters: &[String],
+        namespace: T,
```
Comment (nightkr) [inline src/platform/stack/spec.rs:119]: And here.

File: src/platform/stack/spec.rs [hunk 14]
```diff
@@ -77,41 +106,114 @@ impl StackSpecV2 {
...
 
     #[instrument(skip_all)]
-    pub fn install_demo_manifests(
+    pub async fn install_demo_manifests<T>(
         &self,
+        manifests: &Vec<ManifestSpec>,
         valid_demo_parameters: &[DemoParameter],
         demo_parameters: &[String],
-    ) -> Result<(), StackError> {
+        namespace: T,
```
Comment (nightkr) [inline src/platform/stack/spec.rs:141]: And here.

File: src/platform/stack/spec.rs [hunk 15]
```diff
@@ -77,41 +106,114 @@ impl StackSpecV2 {
...
+        Self::install_manifests(manifests, &parameters, namespace).await?;
+        Ok(())
+    }
+
+    /// Installs a list of templated manifests inside a namespace.
+    #[instrument(skip_all)]
+    async fn install_manifests<T>(
+        manifests: &Vec<ManifestSpec>,
+        parameters: &HashMap<String, String>,
+        namespace: T,
```
Comment (nightkr) [inline src/platform/stack/spec.rs:162]: And here

File: src/utils/check.rs [hunk 16]
```diff
@@ -8,11 +9,14 @@ pub fn binary_present<T: AsRef<OsStr>>(name: T) -> bool {
 }
 
 /// Returns if ALL binaries in the list are present in the $PATH.
+#[instrument]
 pub fn binaries_present<T, L>(names: L) -> bool
```
Comment (nightkr) [inline src/utils/check.rs:13]: This could be simplified to `(names: &[impl AsRef<OsStr>])`

File: src/utils/mod.rs [hunk 17]
```diff
@@ -2,3 +2,11 @@ pub mod check;
 pub mod params;
 pub mod path;
 pub mod read;
+
+/// Returns the name of the operator used in the Helm repository.
+pub fn operator_name<T>(name: T) -> String
```
Comment (nightkr) [inline src/utils/mod.rs:7]: Also here.

Comment (nightkr) [reply src/utils/mod.rs:7]: I'd also consider naming it something like `operator_chart_name`.. the current name is quite ambiguous.

File: bins/stackablectl/src/cmds/operator.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs]: Changed in 106a9d8.

File: bins/stackablectl/src/cmds/operator.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nightkr) [reply bins/stackablectl/src/cmds/operator.rs]: Mostly the implementation side.. something like:

```
trait DisplayTable {
    fn columns() -> &'static [(String, fn(&Self) -> String)];
}

fn print_command_output(mode: OutputMode, out: impl Serialize + DisplayTable) -> String {
    match mode {
        OutputMode::Table => ...,
        OutputMode::Json => ...,
        OutputMode::Yaml => ...,
    }
}
```

But again, doesn't have to be now.

File: bins/stackablectl/src/cmds/operator.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs]: Removed in a538ac2.

File: bins/stackablectl/src/cmds/operator.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (nightkr) [reply bins/stackablectl/src/cmds/operator.rs:193]: I'd rather remove it for now.

File: src/utils/mod.rs (hunk 17 above)
Comment (Techassi) [reply src/utils/mod.rs]: Changed in df0076f.

File: src/utils/check.rs (hunk 16 above)
Comment (Techassi) [reply src/utils/check.rs]: Changed in df0076f.

File: src/platform/stack/spec.rs (hunk 15 above)
Comment (Techassi) [reply src/platform/stack/spec.rs]: Changed in df0076f.

File: src/platform/stack/spec.rs (hunk 14 above)
Comment (Techassi) [reply src/platform/stack/spec.rs]: Changed in df0076f.

File: src/platform/stack/spec.rs (hunk 13 above)
Comment (Techassi) [reply src/platform/stack/spec.rs]: Changed in df0076f.

File: src/platform/stack/spec.rs (hunk 12 above)
Comment (Techassi) [reply src/platform/stack/spec.rs]: Changed in df0076f.

File: src/platform/release/spec.rs (hunk 11 above)
Comment (Techassi) [reply src/platform/release/spec.rs]: Changed in df0076f.

File: src/kube.rs (hunk 10 above)
Comment (Techassi) [reply src/kube.rs]: Changed in df0076f.

File: src/kube.rs (hunk 9 above)
Comment (Techassi) [reply src/kube.rs]: Changed in df0076f.

File: src/helm.rs (hunk 7 above)
Comment (Techassi) [reply src/helm.rs]: Changed in df0076f.

File: src/helm.rs (hunk 6 above)
Comment (Techassi) [reply src/helm.rs]: Changed in df0076f.

File: bins/stackablectl/src/cmds/operator.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs]: Removed in 6f1697e.

File: bins/stackablectl/src/cmds/operator.rs [hunk 18]
```diff
@@ -72,18 +98,37 @@ Use \"stackablectl operator describe <OPERATOR>\" to get available versions for
...
+--cluster-cp-nodes argument. The default number of control plane nodes is '1'.
+So when specifying a total number of nodes of '4', there will be one control
+plane node and three worker nodes.")]
+    cluster_nodes: usize,
+
+    /// Number of control plane nodes in the local cluster
+    #[arg(long, default_value_t = 1)]
+    #[arg(long_help = "Number of control plane nodes in the local cluster
+
+This number must be smaller than --cluster-nodes. If this is not the case,
```
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs:127]: This will now error. Changed in 3170bb5.

File: bins/stackablectl/src/cmds/operator.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Techassi) [reply bins/stackablectl/src/cmds/operator.rs]: I would like to introduce this in a separate PR. Marking as resolved for now.

[the remark this answers is not in the document]
Comment (nightkr) [reply bins/stackablectl/src/cmds/operator.rs]: That's fair.

File: src/cluster/kind/mod.rs [hunk 19]
```diff
@@ -90,14 +88,21 @@ impl KindCluster {
...
-            .ok_or(ClusterError::StdinError)?
-            .write_all(config_string.as_bytes())
-            .context(IoSnafu {})?;
-
-        if let Err(err) = kind_cmd.wait() {
+        let mut stdin = kind_cmd.stdin.take().unwrap();
+        tokio::spawn(async move {
+            stdin.write_all(config_string.as_bytes()).await.unwrap();
+            drop(stdin);
+        });
```
Comment (nightkr) [inline src/cluster/kind/mod.rs:95]: We shouldn't need to spawn a separate task for this, you're already waiting for the process to end below.

**Review by nightkr — APPROVED** [review]
